internals/feature_helpers.py

```diff
@@ -45,6 +46,27 @@ def filter_unlisted(feature_list: list[dict]) -> list[dict]:
   return listed_features
 
 
+def filter_confidential_ids(feature_ids: list[int]) -> list[dict]:
+  """Filters a feature id list to return only the ids of features the user should see."""
+  user = users.get_current_user()
+  visible_ids = []
+  for id in feature_ids:
+    if permissions.can_view_feature_with_id(user, id):
```

Comment:
This seems like it would be too slow because each call to `can_view_feature_with_id()` makes a separate round-trip to NDB.  It would be better to call `can_view_feature()` given that the calling code has already loaded the `FeatureEntry` objects.  I think the functions that call this need to move the filtering logic ahead of the conversion-to-dict step.  That probably means that we should change the cache to hold `FeatureEntry` objects rather than the converted python dicts.  So, that might best be done with a prequel PR for just that change.